File: ExamPrep/ProgrammingQuestionsInAllLanguages/Racket-3/main.js

class BST{

    constructor(){
        this.head = null;
    }

    getHead(){
        return this.head;
    }

    inOrder(node){
        if(node === null){
            return;
        }

        this.inOrder(node.getRight());
        console.log(`${node.getLetter()}: ${node.getFreq()}`);
        this.inOrder(node.getLeft());
    }

    printTree(){
        this.inOrder(this.getHead());
    }

    insertNode(newNode){
        this.head = this.insertNodeHelper(this.getHead(), newNode);

    }
    insertNodeHelper(currNode, newNode){
        if(currNode === null){
            return newNode;
        }
        if(newNode.getLetter() === currNode.getLetter()){
            currNode.increaseFreq();
        }
        else if(newNode.getLetter() > currNode.getLetter()){
            currNode.setLeft(this.insertNodeHelper(currNode.getLeft(), newNode));
        }
            
        else{
            currNode.setRight(this.insertNodeHelper(currNode.getRight(), newNode));
        }
        return currNode;
    }
}

class Node {
    constructor(letter) {
        this.letter = letter;
        this.left = null;
        this.right = null;
        this.freq = 1;
    }

    getFreq() {
        return this.freq;
    }

    increaseFreq() {
        this.freq += 1;
    }

    getLeft() {
        return this.left;
    }
    getRight() {
        return this.right;
    }

    setLeft(newNode) {
        this.left = newNode;
    }
    setRight(newNode) {
        this.right = newNode;
    }

    getLetter() {
        return this.letter;
    }

    setLetter(newLetter) {
        this.letter = newLetter;
    }
}


function word_to_tree(word, tree) {

    for (var letter of word.toLowerCase()) {
        currNode = new Node(letter)
        tree.insertNode(currNode)
    }
}

tree = new BST()
word_to_tree("HelLloOOOO", tree)
tree.printTree()

File: ExamPrep/ProgrammingQuestionsInAllLanguages/Racket-3/main.py




class BST:
    def __init__(self, head=None):
        self.head = head
    
    def getHead(self):
        return self.head
    
    def inOrder(self, node):
        if(node is None):
            return
        
        self.inOrder(node.getLeft())
        print(f'{node.getLetter()}: {node.getFreq()}')
        self.inOrder(node.getRight())
    
    def printTree(self):
        self.inOrder(self.getHead())
    
    def insertNode(self, newNode):
        self.head = self.insertNodeHelper(self.getHead(), newNode)
        
    def insertNodeHelper(self, currNode, newNode):
        
        if(currNode is None):
            return newNode
        
        if(newNode.getLetter() == currNode.getLetter()):
            currNode.increaseFreq()
            
        elif(newNode.getLetter() < currNode.getLetter()):
            currNode.setLeft(self.insertNodeHelper(currNode.getLeft(), newNode))
            
        else:
            currNode.setRight(self.insertNodeHelper(currNode.getRight(), newNode))
        
        return currNode

   

class Node:
    def __init__(self, left = None, right = None, letter = "", freq = 1):
        self.left = left
        self.right = right
        self.letter = letter
        self.freq = freq
        
    def getFreq(self):
        return self.freq
    
    def increaseFreq(self):
        self.freq += 1
        
    def getRight(self):
        return self.right 
    
    def getLeft(self):
        return self.left
    
    def setLeft(self, newNode):
        self.left = newNode
        
    def setRight(self, newNode):
        self.right = newNode
    
    def getLetter(self):
        return self.letter
    
    def setLetter(self, newLetter):
        self.letter = newLetter
            

def word_to_tree(word, tree):
    for curr in word.lower():
        node = Node(letter = curr)
        tree.insertNode(node)
        
tree = BST()
word_to_tree("HelLloOOOO", tree)
tree.printTree()
        